### ISSUE TITLE
Supporting Nori's IPLD use case

### ISSUE THREAD

**Issue by pcardune** [ipld/ipld#28]
tl;dr: After discussing some of this with @diasdavid and @vmx, they suggested we post an issue here for further conversation. I'm hoping to get some feedback/guidance on our approach in the comments! Thanks in advance to anyone who can help us out 😃 

# Nori's IPLD Use Case

## High level

We are creating a commodity token called a "Carbon Removal Credit" (CRC) on the
ethereum blockchain which represents a certain amount of carbon having been
removed. The data collected during the removal process serves as proof of
removal and so each CRC will be linked to the data that proves it's
legit. Storing all that data in the Ethereum blockchain is cost prohibitive, so
IPFS, IPLD, and eventually FileCoin, become interesting mechanisms for
decentralizing access to that public data.

## Example Data

The data we wish to distribute can be represented as a graph of immutable
structured nodes, and binary file data (images and such). As a simple example,
let's say we have various node types included `project`, `plot`, `crc`, `data`,
`validation`. Here are examples of what some of those might look like.

**project node**

This is a node representing a project to remove carbon dioxide from the
atmosphere. We'll call it `project-1`.

```json
{
  "node": {
    "name": "Paul's Farm",
    "location": {
      "city":"Seattle",
      "state": "WA",
    },
    "createdOn": "2018-01-14",
    "url": "http://paulsfarm.com"
  },
  "edges": {
    "owner": ["user-1"]
  }
}

```

**land plot node**

This is a node representing a plot of land belonging to the project owner. We'll
call it `plot-1`,

```json
{
  "node": {
    "name": "East Plot",
    "area": [
      {"lat": 45.2342, "long": -122.4342"},
      {"lat": 45.3342, "long": -122.5342"},
      {"lat": 45.3342, "long": -122.5342"},
      {"lat": 45.2342, "long": -122.4342"},
    ],
  },
  "edges": {
    "project": ["project-1"],
    "createdBy": ["user-1"],
    "historicalData": ["data-1", "data-2", "data-3", "data-4"]
  }
}
```

**CRC Data node**

This is a node with all the info, or links to the info, needed to validate a
certain amount of carbon dioxide having been removed.

```json
{
  "node": {
    "generationDate": "2018-02-14",
    "carbonRemoved": 1.23,
    "removalMethod": "soil",
    "measurements": {
      "complicated": "science stuff",
    }
  },
  "edges": {
    "validationRecords": ["validation-1", "validation-2"],
    "project": ["project-1"],
    "createdBy": ["user-1"],
    "plots": ["plot-1", "plot-2"]
  }
}
```

## How the data will be used

1. Linked from non-fungible ethereum token

   We'll link each newly minted CRC token (just a struct in ethereum) to one of
   the CRC data nodes, for example: `crc-1`. A CID is exactly what we need for this.

1. Fetched from client-side javascript code

   Browser based ethereum clients (metamask/web3) will need to look up the
   `crc-1` data by the identifier stored with the ethereum token, so it can
   display that data on a webpage. It will probably need to fetch many nodes
   linked to from the `crc-1`, such as `project-1`, `plot-1`, and `user-1` at
   the same time and in an efficient way.

1. Queried with a wide range of query parameters from client-side javascript

   Some example queries include:

   * The 100 most recently created `crc` nodes linked to `project` nodes with
     `state="WA"`.

   * The sum of all the `carbonRemoved` fields of all `crc` nodes linked to
     `project` nodes with `state="WA"` where `generationDate` is in the year
     2017

## Scaling

Near term: 10k-99k nodes
Long term: billions of nodes

## How we are doing this now

Centralized cloud database solutions solve all of the above problems except for
the "decentralized" part. So we're just using Google's Cloud DataStore for
everything, but trying to do it in a way that is "compatible" with decentralized
technologies like IPFS:

1. We write individual nodes to Cloud DataStore (basically mongoDB).
1. We serialize the same nodes using `ipld-dag-cbor` and use the CID as the
   primary key when writing to Cloud DataStore.
1. We also store the `dag-cbor` serialized data in a special attribute on the
   Cloud DataStore object.
1. We store binary data (images and such) into Google's Cloud Storage service
   (same as s3) also keyed off the CID of that data.
1. We make the data stored in Cloud DataStore available to client-side
   javascript via a GraphQL API.

This allows us to scale to at least hundreds of millions of nodes (assuming
Cloud DataStore actually works), perform very fast lookups and queries,
easily backup the data into cold storage, and toss it into data warehouses for
more complex analysis.

But of course people will have to rely on us continuing to pay the bills and
operate this centralized service in order to maintain access to this data. And
they will have to trust us to deliver accurate query responses. So we want to
distribute the data and the query indexes to as many people as possible as a way
to secure the data "forever". We could distribute the data by providing links to
gigantic petabyte database backup files which people can theoretically download
and do something with, but IPFS, IPLD, and FileCoin sound like better options :)

## Pathway to decentralization

Ultimately we want people to be able to use the platform we are building without
having to go through us. While that is probably quite a ways off, here are the
steps we could take in that direction:

1. Make the IPLD encoded data in our database available through a public API
   (graphql/rest/whatever) hosted on our servers.
   * Still centralized, but the data is public. CIDs ensure that the
     data we are serving is accurate, so we can't cheat.
2. Mirror/pin the IPLD encoded data in our database to a cluster of IPFS nodes that
   we host.
   * Still centralized, but now it is "easy" for others to pin the data on their
     own ipfs nodes if they want, and make it available to everybody else
     through standard ipfs APIs. Now there is an alternative to our public
     graphql API.
   * If our IPFS cluster goes down, we can repopulate it from the database.
   * Would cost us a lot of extra time and money to manage an IPFS cluster on
     our own. Doable, but not fun until someone comes along with IPFS
     cluster-as-a-service. Then it would just cost us money, but not time.
   * Maybe we can write our own IPFS resolver that points to google cloud datastore directly?
3. Switch our own client-side javascript over to IPFS (instead of our public
   API) for fetching IPLD nodes.
   * If public IPFS latency stays the same as it is today, our code would have
     to fall back to our own public graphql API when IPFS doesn't respond fast
     enough.
4. Release an open source javascript library for interacting with our data over
   IPFS.
   * Theoretically decentralized, assuming other people choose to run their own
     IPFS based mirrors.
   * Now other folks can theoretically build things that use data from our
     platform without relying on our graphql API being available.
5. Release an open source server application that facilitates the mirroring of
   subsets of CRC data.
   * Presumably the people who own CRCs or supply CRCs would be incentivized to
     mirror at least the data associated with CRCs that they own or supplied.
   * Maybe this is enough to be considered fully decentralized, except you still
     don't have decentralized guarantees of data permanence.
6. Use transaction fees or some other decentralized funding mechanism to pay for
   FileCoin to store this data "permanently".
   * Probably about as close as you can get to guaranteed decentrailized
     permanence and availability. Sounds like there are still a lot of hard hard
     problems to solve before this could be a reality.

These steps don't really include anything about decentralized queryability since
I don't really have a good idea of how that would work in practice. But I sure
would like it to happen!

Also, each one of these steps could trigger very long discussions about the various
ways to implement each step. We would certainly like solutions that are low
cost, low latency, high performance, etc. etc. and will look to the community
for guidance on achieving that.

**Comment by vmx** [comment]
Thanks a lot for the detailed write-up. Your pathway to decentralization sounds sensible, though I'd like to get that double-checked by people like @diasdavid who have been longer in this business.

**Comment by diasdavid** [comment]
@hermanjunge would you mind moving your writeup to a separate issue? It will be valuable to have each use case in each issue.

**Comment by diasdavid** [comment]
@hermanjunge given that you moved the Metamask use case to #29 (Thanks!) I remove the two comments from this thread to keep it focused on Noris case :)

**Comment by pcardune** [comment]
@hermanjunge 
Let me see if I understand correctly. Assuming we have an append-only database (like a block chain), which we want to do complex queries on (aggregations, joins, filters, sorts, etc.), then the main thing that would be continuously changing are the query indexes.

So what you are suggesting is that we merklelize, shard, distribute, and update the always changing query indexes.

Since I'm talking about query indexes, I will add another pain point: how do you distribute the query computations to the relevant shards and aggregate their results? Map reduce seems like the right approach, though my understanding is that there are some non-trivial limitations. You either have to deal with some percent of shards not being available and therefore getting only probabilistic results back, or you have to take the performance hit of waiting for the missing shards to become available before returning the final 100% accurate result.

**Comment by hermanjunge** [comment]
@pcardune 

> So what you are suggesting is that we merklelize, shard, distribute, and update the always changing query indexes.

Indeed. Similar to transactions and blocks in the blockchain, take the mutating indexes, sign them into a set of operations to merklelize, recompute the state root of that tree and publish it.

> You either have to deal with some percent of shards not being available and therefore getting only probabilistic results back

In this model you need to have an obscene redundancy (maybe 10x per shard?), a good random criteria (could be based on peer id) to assign the shards, and always a couple of full shard nodes.

Now for the latter point. I am aware your goal is **to do without having to maintain storage**, so we take the idea of "_you keep all the shards in the server_" idea out of the table. One key development here, then, is to **create a good incentivization system for peers to keeping shards available**. In the particular use case for ethereum storage, we are counting with users in the need to maintain cache hubs to boost availability (and of course for `go-ethereum` and `parity` users to enter libp2p someday). In the general use case, until filecoin is not released, some kind of secondary token could be useful to support this storage maybe?

**Comment by pcardune** [comment]
> In the general use case, until filecoin is not released, some kind of secondary token could be useful to support this storage maybe?

We're willing to bare the cost of centralized storage in the short and long term to guarantee availability. We see filecoin more as a way to further remove ourselves as a data intermediary, and less as a cost saving measure. Unless filecoin takes a decade to materialize, I'm not so worried about this part happening in a reasonable timeframe.

That said, I am interested in reducing the cost of the centralized storage, while still making everything available in the p2p network. As the standard IPFS repo is currently implemented, the daemon needs access to a local filesystem to read/write the ipfs blocks. But a local file system is far from the most cost effective or reliable way to store and retrieve data in a public cloud environment. I'd love to see a first class IPFS repo implementation that's backed by and optimized for Google's Cloud Storage and Cloud Datastore for example.

**Comment by hermanjunge** [comment]
> But a local file system is far from the most cost effective or reliable way to store and retrieve data in a public cloud environment

We are currently working on IPLD bridges that will make data available from a connected Redis DB... 

This is a short term solution, so, joining your petition @pcardune, @whyrusleeping Question: How can we help on an effort to mount a different data store into a go-ipfs client?

**Comment by whyrusleeping** [comment]
@hermanjunge we basically just need to add a datastore plugin to go-ipfs. We already have the datastore nicely changeable in the config, it just needs to be able to understand whatever datastore you want to use.

cc @magik6k who has done both lots of plugin work and lots of datastore work.

**Comment by diasdavid** [comment]
@hermanjunge same for js-ipfs, that is how @kumavis got the Parity to IPFS demo to work a while ago. Read the thread https://github.com/ipfs/js-ipfs/issues/763

**Comment by magik6k** [comment]
Datastore plugins would do the job, but I think we could try to generalize filestore a bit to support more pluggable block providers (it is already layered on top of the blockstore).

Blockstore could be pluggable, with default configuration being something like `['datastore', 'filestore']`. This would provide a nicer interface for plugins which just want to implement a block provider than having to implement the whole datastore interface(s).
